linkerd/app/outbound/src/discover.rs

```diff
@@ -68,109 +72,186 @@ pub(crate) fn resolver(
-                            dispatcher,
-                        )
-                    };
-
-                    let policy = policy_from_profile(&*profile_rx.borrow_and_update());
-                    tracing::debug!(
-                        ?policy,
-                        "Policy not found; synthesizing policy from profile..."
-                    );
-                    let (tx, policy_rx) = watch::channel(policy);
-                    tokio::spawn(async move {
-                        loop {
-                            if profile_rx.changed().await.is_err() {
-                                tracing::debug!("Profile watch closed, terminating");
-                                return;
-                            }
-
-                            let policy = policy_from_profile(&*profile_rx.borrow_and_update());
-
-                            tracing::debug!(
-                                ?policy,
-                                "Profile updated; synthesizing policy from profile..."
-                            );
-                            if tx.send(policy).is_err() {
-                                tracing::debug!("Policy watch closed, terminating");
-                                return;
-                            }
-                        }
-                    });
-                    return Ok((Some(profile), policy_rx));
-                }
-
-                let policy = policy::ClientPolicy::from_backend(
+            // If there was a profile resolution, try to use it to synthesize a
+            // enpdoint policy.
+            if let Some(profile) = profile {
+                let policy = spawn_synthesized_profile_policy(
+                    orig_dst,
+                    profile.clone().into(),
+                    queue,
                     detect_timeout,
-                    default_queue,
-                    policy::BackendDispatcher::Forward(addr.0, Default::default()),
-                );
-                tracing::debug!(
-                    ?policy,
-                    "Policy not found; no profile: synthesizing policy from original destination address"
                 );
-
-                let (tx, policy_rx) = watch::channel(policy);
-                // keep the watch channel alive as long as it's in use.
-                tokio::spawn(async move { tx.closed().await });
-                return Ok((None, policy_rx));
+                return Ok((Some(profile), policy));
             }
 
-            // Otherwise, errors resolving the policy are considered fatal.
-            Err(policy_error)
+            // Otherwise, route the request to the original destination address.
+            let policy = spawn_synthesized_origdst_policy(orig_dst, queue, detect_timeout);
+            Ok((None, policy))
         })
     })
 }
 
+#[inline]
+fn is_not_found(e: &Error) -> bool {
+    e.downcast_ref::<tonic::Status>()
+        .map(|s| s.code() == tonic::Code::NotFound)
+        .unwrap_or(false)
```

Comment:
currently the error is always a `Status` at the top level, but it occurs to me that this is a bit brittle in that the error might get wrapped...maybe this should traverse a whole source chain if there is one? we can address separately though.

Reply:
oh yeah we should use the `errors` helper here probably but it's fine for now.